echopype/echodata/simrad.py

```diff
@@ -120,32 +120,28 @@ def _retrieve_correct_beam_group_EK80(
     power_ed_group = None
     complex_ed_group = None
 
-    transmit_type_has_BB = not np.all(echodata["Sonar/Beam_group1"]["transmit_type"] == "CW")
+    transmit_type = echodata["Sonar/Beam_group1"]["transmit_type"]
+    # assume transmit_type identical for all pings in a channel
+    first_ping_transmit_type = transmit_type.isel(ping_time=0)
     if waveform_mode == "BB":
         # check BB waveform_mode, BB must always have complex data, can have 2 beam groups
-        # when echodata contains CW power and BB complex samples, and transmit_frequency_start
-        # variable in Beam_group1
-        if not transmit_type_has_BB:
-            raise ValueError("waveform_mode='BB', but broadband data not found!")
-        elif "backscatter_i" not in echodata["Sonar/Beam_group1"].variables:
+        # when echodata contains CW power and BB complex samples
+        if np.all(first_ping_transmit_type == "CW"):
             raise ValueError("waveform_mode='BB', but complex data does not exist!")
         elif echodata["Sonar/Beam_group2"] is not None:
             power_ed_group = "Sonar/Beam_group2"
             complex_ed_group = "Sonar/Beam_group1"
         else:
             complex_ed_group = "Sonar/Beam_group1"
-
     else:
         # CW can have complex or power data, so we just need to make sure that
         # 1) complex samples always exist in Sonar/Beam_group1
         # 2) power samples are in Sonar/Beam_group1 if only one beam group exists
         # 3) power samples are in Sonar/Beam_group2 if two beam groups exist
 
-        # Raise error if waveform_mode="CW" but CW data does not exist
-        if encode_mode == "complex" and transmit_type_has_BB:
-            # complex + BB data (all channels are BB)
-            if np.all(echodata["Sonar/Beam_group1"]["transmit_type"] != "CW"):
-                raise ValueError("waveform_mode='CW', but all data are broadband (BB)!")
+        # Raise error if waveform_mode="CW" but CW data does not exist (not a single ping is CW)
+        if encode_mode == "complex" and np.all(transmit_type != "CW"):
+            raise ValueError("waveform_mode='CW', but all data are broadband (BB)!")
```

Comment:
Since we only test for `first_ping_transmit_type` in the BB case, to make things more clear when we pick up the issue of within-channel CW-BB switches, maybe we change the check below to just the first ping too?
```suggestion
        # Raise error if waveform_mode="CW" but CW data does not exist (not a single ping is CW)
        # TODO: this needs to be changed when fixing within-channel CW-BB switch
        if encode_mode == "complex" and np.all(first_ping_transmit_type != "CW"):
            raise ValueError("waveform_mode='CW', but all data are broadband (BB)!")
```

Reply:
Looks good. I'll accept the suggestion here rather than via a push from me.